include/alpaka/core/RuntimeMacros.hpp

```diff
@@ -0,0 +1,48 @@
+/* Copyright 2022  Andrea Bocci, Mehmet Yusufoglu, René Widera, Aurora Perego
+ * SPDX-License-Identifier: MPL-2.0
+ */
+
+#pragma once
+
+// Implementation details
+#include "alpaka/core/Sycl.hpp"
+
+//! ALPAKA_DEVICE_THROW either aborts(terminating the program and creating a core dump) or throws std::runtime_error
+//! depending on the Acc. The std::runtime_error exception can be catched in the catch block.
+//!
+//! CUDA __trap function triggers std::runtime_error but can be catched during wait not exec.
+//!
+//! The OpenMP specification mandates that exceptions thrown by some thread must be handled by the same thread.
+//! Therefore std::runtime_error thrown by ALPAKA_DEVICE_THROW aborts the the program for OpenMP backends. If needed
+//! the SIGABRT signal can be catched by signal handler.
+#if defined(ALPAKA_ACC_GPU_CUDA_ENABLED) && defined(__CUDA_ARCH__)
+#    define ALPAKA_DEVICE_THROW(MSG)                                                                                  \
+        {                                                                                                             \
+            printf(                                                                                                   \
+                "alpaka encountered a user-defined error condition while running on the CUDA back-end:\n%s",          \
+                (MSG));                                                                                               \
+            __trap();                                                                                                 \
+        }
+#elif defined(ALPAKA_ACC_GPU_HIP_ENABLED) && defined(__HIP_DEVICE_COMPILE__)
+#    define ALPAKA_DEVICE_THROW(MSG)                                                                                  \
+        {                                                                                                             \
+            printf(                                                                                                   \
+                "alpaka encountered a user-defined error condition while running on the HIP back-end:\n%s",           \
+                (MSG));                                                                                               \
+            abort();                                                                                                  \
```

Comment:
@mehmetyusufoglu reported offline that aborting is not working. We use `__builtin_trap();` what is working fine.

Reply:
`__builtin_trap();`  halts the main thread (host) but does not exit from the program.  

The only working command seems to be `assert(fail)`. I used assert(fail) because the code below aborted with a core dump.  CAn not be catched as an std::Excception or hipError.



`#include <hip/hip_runtime.h>
#include <iostream>
#include <signal.h>

// Kernel definition
__global__ void myKernel(int *data, int threshold) {
  int idx = blockIdx.x * blockDim.x + threadIdx.x;

  // Trigger a segmentation fault by accessing invalid memory
  // asm volatile("s_trap 0x1;");
  //            __builtin_trap();a
  if (idx == 0) {
    printf("Aborting kernel at index %d\n", idx);

    assert(false);
  }
  // Kernel operation
  data[idx] *= 2;
}

int main() {
  const int dataSize = 10;
  int h_data[dataSize] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
  int threshold = 7;

  int *d_data;
  hipMalloc(&d_data, dataSize * sizeof(int));
  hipMemcpy(d_data, h_data, dataSize * sizeof(int), hipMemcpyHostToDevice);

  dim3 blockDim(5);
  dim3 gridDim((dataSize + blockDim.x - 1) / blockDim.x);
  try {
    // Launch the kernel
    hipLaunchKernelGGL(myKernel, gridDim, blockDim, 0, 0, d_data, threshold);
    // hipDeviceSynchronize();
    // Wait for GPU to finish before accessing on host
  } catch (...) {
    printf("catched from try");
  }
  hipError_t err = hipDeviceSynchronize();
  //  Check for errors
  if (err != hipSuccess) {
    std::cerr << "Kernel execution aborted or encountered an error: "
              << hipGetErrorString(err) << std::endl;
    std::exit(EXIT_FAILURE);
  }
  printf("copy data back to host:\n");

  // Copy data back to host
  //  hipMemcpy(h_data, d_data, dataSize * sizeof(int), hipMemcpyDeviceToHost);

  // Print the results
  //    for (int i = 0; i < dataSize; i++) {
  //      std::cout << h_data[i] << " ";
  // }
  std::cout << std::endl;

  // Cleanup
  hipFree(d_data);

  return 0;
}
// compiled by command: hipcc kernel_abort.cpp -o kernel_abort
// rocm version: rocm-5.7.2
`